ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/StateCachesTest.java

```diff
@@ -0,0 +1,171 @@
+  }
+
+  @Test
+  void SlotCaches_processSlotShouldCleanUpSlotCaches()
+      throws SlotProcessingException, EpochProcessingException {
+
+    BeaconStateCache.getSlotCaches(stateWithCaches)
+        .increaseBlockProposerRewards(UInt64.valueOf(84));
+    BeaconStateCache.getSlotCaches(stateWithCaches).setBlockExecutionValue(UInt256.valueOf(42));
+
+    final BeaconState stateAtSlot3 = spec.processSlots(stateWithCaches, UInt64.valueOf(3));
+
+    assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3).getBlockExecutionValue())
+        .isEqualByComparingTo(UInt256.ZERO);
+    assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3).getBlockProposerRewards())
+        .isEqualByComparingTo(UInt64.ZERO);
+  }
+
+  @ParameterizedTest
+  @EnumSource(BlockRewardsSource.class)
+  void SlotCaches_blockProcessingShouldIncreaseBlockRewardsWhenBlockHasAttestations(
+      final BlockRewardsSource blockRewardsSource)
+      throws SlotProcessingException, EpochProcessingException, BlockProcessingException {
+
+    assertThat(BeaconStateCache.getSlotCaches(stateWithCaches).getBlockExecutionValue())
+        .isEqualByComparingTo(UInt256.ZERO);
+    assertThat(BeaconStateCache.getSlotCaches(stateWithCaches).getBlockProposerRewards())
+        .isEqualByComparingTo(UInt64.ZERO);
+
+    final BeaconState stateAtSlot3 = spec.processSlots(stateWithCaches, UInt64.valueOf(3));
+
+    assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3).getBlockExecutionValue())
+        .isEqualByComparingTo(UInt256.ZERO);
+    assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3).getBlockProposerRewards())
+        .isEqualByComparingTo(UInt64.ZERO);
+
+    // execution value should not be affected by block processing
+    BeaconStateCache.getSlotCaches(stateAtSlot3).setBlockExecutionValue(UInt256.valueOf(42));
+
+    final SignedBeaconBlock blockAtSlot3 = createWorthyBlock(blockRewardsSource);
+
+    final BeaconState stateAtSlot3WithBlock =
+        spec.atSlot(UInt64.valueOf(3))
+            .getBlockProcessor()
+            .processUnsignedBlock(
+                stateAtSlot3,
+                blockAtSlot3.getMessage(),
+                IndexedAttestationCache.NOOP,
+                BLSSignatureVerifier.NO_OP,
+                Optional.empty());
+
+    if (blockRewardsSource == BlockRewardsSource.NONE) {
+      assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3WithBlock).getBlockProposerRewards())
+          .isEqualByComparingTo(UInt64.ZERO);
+    } else {
+      assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3WithBlock).getBlockProposerRewards())
+          .matches(rewards -> rewards.isGreaterThan(UInt64.ZERO));
+    }
+
+    // execution value should not be affected by block processing
+    assertThat(BeaconStateCache.getSlotCaches(stateAtSlot3WithBlock).getBlockExecutionValue())
+        .isEqualByComparingTo(UInt256.valueOf(42));
+  }
+
+  private SignedBeaconBlock createWorthyBlock(final BlockRewardsSource blockRewardsSource) {
```

Comment:
it would be great to also make a test that we got sum of this if all are in block if it doesn't look too complex

Reply:
problem is that unless i calculate the exact expected reward, I cannot differentiate between one or more rewards

Reply:
maybe I can try to reuse `RewardCalculator` if I don't fall into dependency hell

Reply:
> problem is that unless i calculate the exact expected reward, I cannot differentiate between one or more rewards
I meant run all separate first, then together